imap_processing/codice/codice_l1a_de.py

```diff
@@ -0,0 +1,468 @@
+        epoch_start = packet_index_starts[epoch_index] * num_priorities
+        epoch_end = packet_index_starts[epoch_index + 1] * num_priorities
+        # Extract the decompressed data for current epoch.
+        # epoch_data should be of shape ((num_priorities * num_events),)
+        epoch_data = decompressed_data[epoch_start:epoch_end]
+
+        # Extract these other data
+        unordered_priority = priorities_arr[epoch_start:epoch_end]
+        unordered_data_quality = data_quality_arr[epoch_start:epoch_end]
+        un_ordered_num_events = num_events_arr[epoch_start:epoch_end]
+
+        # If priority array unique size is not same size as
+        # num_priorities, then throw error. They should match.
+        if len(np.unique(unordered_priority)) != num_priorities:
+            raise ValueError(
+                f"Priority array for epoch {epoch_index} contains "
+                f"non-unique values: {unordered_priority}"
+            )
+
+        # Until here, we have the out of order priority data. Data could have been
+        # collected in any priority order. Eg.
+        #   priority - [0, 4, 5, 1, 3, 2, 6, 7]
+        # Now, we need to put data into their respective priority indexes
+        # in final arrays for the current epoch. Eg. put data into
+        #   priority - [0, 1, 2, 3, 4, 5, 6, 7]
+        de_data["num_events"][epoch_index, unordered_priority] = un_ordered_num_events
+        de_data["data_quality"][epoch_index, unordered_priority] = (
+            unordered_data_quality
+        )
+
+        # Fill the event data into it's bin in same logic as above. But
+        # since the epoch has different num_events per priority,
+        # we need to loop and index accordingly. Otherwise, numpy throws
+        # 'The detected shape was (n,) + inhomogeneous part' error.
+        for priority_index in range(len(unordered_priority)):
+            # Get num_events
+            priority_num_events = int(un_ordered_num_events[priority_index])
+            # Reshape epoch data into (num_events, 8). That 8 is 8-bytes that
+            # make up 64-bits. Therefore, combine last 8 dimension into one to
+            # get 64-bits event data that we need to unpack later.
+            events_in_bytes = np.array(epoch_data[priority_index]).reshape(
+                priority_num_events, 8
+            )
+
+            # combine last 8 dimension into one 64-bits value
+            combined_64bits = np.dot(events_in_bytes, 1 << (8 * np.arange(8)[::-1]))
+            # Put event data into their respective priority number bins
+            priority_num = int(unordered_priority[priority_index])
+            de_data["64bits_event_data"][
+                epoch_index, priority_num, :priority_num_events
+            ] = combined_64bits
+
+    # Now unpack the 64-bits event data using bitwise operations
+    unpacked_vars = list(bit_structure.keys())
+    unpacked_bits = [bit_structure[field]["bit_length"] for field in bit_structure]
```

Comment:
Combine these into a single mapping {field: num_bits} that you can pass in as one value and iterate over within the unpack_bits function.
```suggestion
    unpacked_bits = {field: bit_structure[field]["bit_length"] for field in bit_structure}
```